### PR TITLE
Rules for optimizing idempotency for bitwise operations.

### LINKED ISSUE #8670: Optimiser confused by something
This is the most minimal example I've got so far:
```
contract C {
  function f() external {
        bytes memory amount;
        bytes24 zero_bytes24;
        bytes32 signature_root;
        bytes32 node = sha256(abi.encodePacked(
            sha256(abi.encodePacked(amount, zero_bytes24, signature_root))
        ));
  }
}
```

With `--optimize` (default settings, albeit tried with `--optimize-runs 10000000` too) there is this snippet in the output:
```
        /* "optim.sol":197:251  abi.encodePacked(amount, zero_bytes24, signature_root) */
      pop
      pop
      pop
      swap1
      pop
      add
      dup4
      not(0xffffffffffffffff)
      and
      not(0xffffffffffffffff)
      and
      dup2
      mstore
      0x18
      add
```

The `not and not and` part doesn't seem to make any sense.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #8670.

### CODE REVIEW

File: test/libyul/yulOptimizerTests/fullSimplify/operations.yul [hunk 1]
```diff
@@ -37,8 +38,9 @@
 //     mstore(12, 0)
 //     mstore(13, 0)
 //     mstore(14, 0)
-//     mstore(15, 0)
-//     mstore(16, 0)
-//     mstore(17, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
-//     mstore(18, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
+//     let _39 := not(x)
+//     mstore(15, and(x, _39))
```
Comment (axic) [inline test/libyul/yulOptimizerTests/fullSimplify/operations.yul]: `and(x, not(x))` seems like a convoluted way to write `0` 😉

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSimplify/operations.yul]: weird, that should also be optimized away :(

Comment (chriseth) [reply test/libyul/yulOptimizerTests/fullSimplify/operations.yul]: I guess it is the order of operations now.

File: libevmasm/RuleList.h [hunk 2]
```diff
@@ -662,7 +684,7 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleList(
 	rules += simplificationRuleListPart1(A, B, C, W, X);
 	rules += simplificationRuleListPart2(A, B, C, W, X);
 	rules += simplificationRuleListPart3(A, B, C, W, X);
-	rules += simplificationRuleListPart4(A, B, C, W, X);
```
Comment (chriseth) [inline libevmasm/RuleList.h:665]: Oh wow! I **removed** part4... no wonder it makes the existing tests worse...

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `51b37c9498265965f63065b10d287332e299d192`:
```
Error: Trailing whitespace found:
 test/libyul/yulOptimizerTests/fullSimplify/operations.yul:24:// 

```
Please check that your changes are working as intended.

File: libevmasm/RuleList.h [hunk 3]
```diff
@@ -661,8 +683,9 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleList(
 	std::vector<SimplificationRule<Pattern>> rules;
 	rules += simplificationRuleListPart1(A, B, C, W, X);
 	rules += simplificationRuleListPart2(A, B, C, W, X);
-	rules += simplificationRuleListPart3(A, B, C, W, X);
 	rules += simplificationRuleListPart4(A, B, C, W, X);
+	rules += simplificationRuleListPart4_5(A, B, C, W, X);
+	rules += simplificationRuleListPart3(A, B, C, W, X);
```
Comment (leonardoalt) [inline libevmasm/RuleList.h]: Why is 3 after 4 now?

File: libevmasm/RuleList.h [hunk 4]
```diff
@@ -232,6 +232,28 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart4(
...
+{
+	using Builtins = typename Pattern::Builtins;
+	return std::vector<SimplificationRule<Pattern>>{
+		// idempotent operations
+		{Builtins::AND(Builtins::AND(X, Y), Y), [=]{ return Builtins::AND(X, Y); }, true},
+		{Builtins::AND(Y, Builtins::AND(X, Y)), [=]{ return Builtins::AND(X, Y); }, true},
+		{Builtins::AND(Builtins::AND(Y, X), Y), [=]{ return Builtins::AND(Y, X); }, true},
+		{Builtins::AND(Y, Builtins::AND(Y, X)), [=]{ return Builtins::AND(Y, X); }, true},
+		{Builtins::OR(Builtins::OR(X, Y), Y), [=]{ return Builtins::OR(X, Y); }, true},
+		{Builtins::OR(Y, Builtins::OR(X, Y)), [=]{ return Builtins::OR(X, Y); }, true},
```
Comment (bshastry) [inline libevmasm/RuleList.h:252]: Shouldn't there be a rule for `or(or(x,y),x)` and `or(x, or(x,y))` being simplified to `or(x,y)`? Doesn't order in simplification pattern matter?

Comment (bshastry) [reply libevmasm/RuleList.h:252]: oops, ignore previous comment

File: libevmasm/RuleList.h (hunk 3 above)
Comment (chriseth) [reply libevmasm/RuleList.h]: argh!

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**